### PR TITLE
Support nonexact matches for no_proxy

### PR DESCRIPTION
**Pull request by gh123man**
### What does this PR do?

At the moment, `no_proxy` in `datadog.yaml` only supports exact matches. Most software have rules to match domains partially as well. This is useful for example to match all `datadoghq.com` subdomains without having to explicitly list them all down in `no_proxy`.

Issue tracking this: https://github.com/DataDog/datadog-agent/issues/4739

This PR introduces the option to use the go `http` built in proxy function instead of our custom one with a slight modification - instead of only using the environment vars, it uses the settings from the datadog config. The documentation for how go matches proxy URL's can be found here: https://godoc.org/golang.org/x/net/http/httpproxy#Config

If the new setting `no_proxy_nonexact_match` is disabled, a warning will be logged each time a new url is seen that will have different behavior in the future. 

This information is also presented in the status page as well as high level metadata being reported in the metatdata payload. 

### Describe your test plan

Tested by using a local proxy as well as using an invalid proxy. 
With an invalid proxy - the agent logs make it clear if the URL was attempted to be proxied or not.

### CODE REVIEW

File: pkg/util/http/transport.go [hunk 1]
```diff
@@ -57,46 +68,75 @@ func CreateHTTPTransport() *http.Transport {
 // GetProxyTransportFunc return a proxy function for a http.Transport that
 // would return the right proxy depending on the configuration.
 func GetProxyTransportFunc(p *config.Proxy) func(*http.Request) (*url.URL, error) {
```
Comment (hush-hush) [inline pkg/util/http/transport.go:70]: Could we extend the existing test to test nonexect proxy matches (not necessary the `httpproxy` package but the fact it's correctly used).

File: pkg/metadata/host/payload.go [hunk 2]
```diff
@@ -63,4 +69,5 @@ type Payload struct {
 	NetworkMeta   *NetworkMeta      `json:"network"`
 	LogsMeta      *LogsMeta         `json:"logs"`
 	InstallMethod *InstallMethod    `json:"install-method"`
+	ProxyMeta     *ProxyMeta        `json:"proxy-info"`
```
Comment (hush-hush) [inline pkg/metadata/host/payload.go:72]: Could we add tests for this too ?

**Review by hush-hush — CHANGES_REQUESTED** [review]
Aren't we missing some code in `pkg/config` to define the new `no_proxy_nonexact_match` configuration option ?

Also, do we need that option ? if it's enabled by default old agent shouldn't be affected.

File: pkg/util/http/transport.go [hunk 3]
```diff
@@ -57,46 +68,75 @@ func CreateHTTPTransport() *http.Transport {
...
+			return nil, nil
+		}(r)
+
+		// Print a warning if the proxy behavior would change if the new no_proxy behavior would be enabled
+		newURL, _ := proxyConfig.ProxyFunc()(r.URL)
+		if url != newURL && url != nil {
+			urlString := r.URL.String()
+			NoProxyWarningMapMutex.Lock()
+			if _, ok := NoProxyWarningMap[urlString]; !ok {
+				NoProxyWarningMap[SanitizeURL(r.URL.String())] = true
```
Comment (hush-hush) [inline pkg/util/http/transport.go:134]: Could URL contains sensitive data (user/password, and argument). We don't want to those to be sent to our backend. We might want to only to use the scheme and domain.

**Review by hush-hush — CHANGES_REQUESTED** [review]

File: pkg/util/http/transport.go (hunk 3 above)
Comment (gh123man) [reply pkg/util/http/transport.go:134]: I corrected the log lines so they are safer as you suggested.

**Comment by gh123man** [comment]
@hush-hush 
> Aren't we missing some code in pkg/config to define the new no_proxy_nonexact_match configuration option ?
Fixed!

> Also, do we need that option ? if it's enabled by default old agent shouldn't be affected.
The new behavior is currently disabled and it is opt-in (with a deprecation warning log line). Sometime in the future it will be made default and opt-out.

File: pkg/config/config.go [hunk 4]
```diff
@@ -671,6 +671,7 @@ func InitConfig(config Config) {
 	config.SetKnown("proxy.http")
 	config.SetKnown("proxy.https")
 	config.SetKnown("proxy.no_proxy")
+	config.SetKnown("proxy.no_proxy_nonexact_match")
```
Comment (hush-hush) [inline pkg/config/config.go:674]: We need to use `BindEnvAndSetDefault` so the env var `DD_PROXY_NO_PROXY_NONEXACT_MATCH` is supported.

**Review by hush-hush — CHANGES_REQUESTED** [review]
Almost ready to go

File: pkg/config/config.go [hunk 5]
```diff
@@ -202,6 +202,7 @@ func InitConfig(config Config) {
 	config.BindEnvAndSetDefault("tracemalloc_whitelist", "") // deprecated
 	config.BindEnvAndSetDefault("tracemalloc_blacklist", "") // deprecated
 	config.BindEnvAndSetDefault("run_path", defaultRunPath)
+	config.BindEnvAndSetDefault("no_proxy_nonexact_match", false)
```
Comment (hush-hush) [inline pkg/config/config.go:205]: Let's add it to the config YAML template

Comment (gh123man) [reply pkg/config/config.go:205]: right

File: cmd/agent/gui/views/templates/generalStatus.tmpl [hunk 6]
```diff
@@ -102,6 +102,21 @@
     </span>
   </div>
 
+  {{- if .TransportWarnings }}
+    <div class="stat">
+      <span class="stat_title">Transport Proxy Warnings</span>
+      <span class="stat_data">
+      The following URLs used a proxy - but will ignore the proxy in future agent versions via the no_proxy setting.
```
Comment (ruthnaebeck) [inline cmd/agent/gui/views/templates/generalStatus.tmpl:109]:
```suggestion
      The following URLs used a proxy - but will ignore the proxy in future Agent versions with the no_proxy setting. 
```

File: pkg/status/templates/header.tmpl [hunk 7]
```diff
@@ -85,6 +85,18 @@ NOTE: Changes made to this template should be reflected on the following templat
     {{ $key }}: {{ $value }}
   {{- end }}
 
+  {{- if .TransportWarnings }}
+
+  Transport Proxy Warnings
+  ========================
+    The following URLs used a proxy - but will ignore the proxy in future agent versions via the no_proxy setting.
```
Comment (ruthnaebeck) [inline pkg/status/templates/header.tmpl:92]:
```suggestion
    The following URLs used a proxy - but will ignore the proxy in future Agent versions with the no_proxy setting. 
```

**Review by ruthnaebeck — CHANGES_REQUESTED** [review]
Doc's review

**Review by hush-hush — APPROVED** [review]

**Pull request merged**